Collaborative tags new category work

sharidas opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
Adding a new tag which is same as restricted tag except it is not editable by users.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Adding a new tag which is same as restricted tag except it is not editable by users.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Create a new tag `Editable` with name lets say , `editableTag`, by enabling systemtag_management app by navigating to admin page -> `Workflow & tags`
- Create groups `group1`, `group2` under users page ( by enabling user_management app)
- Create users `user1`, `user2`. Add them to `group1`
- Create user `user3` and add it to `group2`.
- Now try to update the `editableTag` in the `Workflow & tags` page, by adding 2 groups `group1` and `group2`.
- Login to `user1`, now the user will see that `editableTag` is editable. It can also be assigned/unassigned.
- Login to `user3`, now the user will see that `editableTag` is not editable. Its only assignable/unassignable.
- Similarly create `visibleTag` under `Visible`, `invisibleTag` under `Invisible`, and `restrictedTag` under `Restricted`. The `editableTag` behaves the same way as it was before.
- Add `group1` to `restrictedTag`.
- Now when logged in as `user3` , `restrictedTag` is not visible. Only `visibleTag` and `editableTag` are visible.


## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Requires: https://github.com/owncloud/systemtags_management/pull/30

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

PVince81 commented on apps/dav/lib/SystemTag/SystemTagPlugin.php:
```diff
@@ -181,6 +183,10 @@ private function createTag($data, $contentType = 'application/json') {
 		}
 
+		if (isset($data['userUneditable'])) {
+			$userUneditable = (bool)$data['userUneditable'];
```
"uneditable" ?? please find a better name or reverse the logic

sharidas replied:
Guess readonly might fit. Because we are preventing edit for regular users.

PVince81 replied:
I guess at this point there is no chance to make it all positive permissions ? "assignable" and "editable" instead of mixing positive with negative ?

sharidas replied:
Updated with the "editable". Logic is adjusted.

PVince81 commented on core/js/systemtags/systemtagmodel.js:
```diff
@@ -39,16 +40,24 @@
 		parse: function(data) {
+			var userEditable;
+			if (data.userUneditable !== undefined) {
+				userEditable = data.userUneditable === false || data.userUneditable === 'false';
```
why is there userUneditable and also userEditable ?

sharidas replied:
Made consistency.

PVince81 commented on lib/public/SystemTag/ISystemTagManager.php:
```diff
@@ -67,14 +67,15 @@ public function getTag($tagName, $userVisible, $userAssignable);
 	 * @since 9.0.0
 	 */
-	public function createTag($tagName, $userVisible, $userAssignable);
+	public function createTag($tagName, $userVisible, $userAssignable, $userUneditable);
```
add a default value as this would break signature and backward compatibilitiy for consumers of this API

sharidas replied:
Done.

PVince81 commented on lib/public/SystemTag/ISystemTagManager.php:
```diff
@@ -102,7 +104,7 @@ public function getAllTags($visibilityFilter = null, $nameSearchPattern = null);
 	 * @since 9.0.0
 	 */
-	public function updateTag($tagId, $newName, $userVisible, $userAssignable);
+	public function updateTag($tagId, $newName, $userVisible, $userEditable, $userAssignable);
```
same here

PVince81 commented on settings/js/settings.js:
```diff
@@ -44,7 +44,7 @@ OC.Settings = _.extend(OC.Settings, {
 					}
 					$.ajax({
-						url: OC.generateUrl('/settings/users/groups'),
+						url: OC.generateUrl('/apps/user_management/groups'),
```
core should not call an API in an apps, we need to solve this

sharidas replied:
Removed this change.

PVince81 reviewed (commented):
I know this is WIP, added a few comments

sharidas committed:
new tag editable added to systemtag

Added new tag editable to the systemtag.
This tag could be editable only to admins
and to the groups who have permission to edit.
For all other users, this tag could be used
to assign/unassign.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented on core/js/systemtags/systemtagmodel.js:
```diff
@@ -15,7 +15,9 @@
 		PROPERTY_USERVISIBLE:	'{' + OC.Files.Client.NS_OWNCLOUD + '}user-visible',
+		PROPERTY_USEREDITABLE:'{' + OC.Files.Client.NS_OWNCLOUD + '}user-editable',
 		PROPERTY_USERASSIGNABLE:'{' + OC.Files.Client.NS_OWNCLOUD + '}user-assignable',
+		PROPERTY_WHITELISTEDINGROUP:'{' + OC.Files.Client.NS_OWNCLOUD + '}whiteListedInGroup'
```
inconsistent property name: all other properties have a dash in them but you used camel case

the name "white listed in group" isn't clear compared to "group", should it be called something like "editableInGroup" instead ?

PVince81 commented on core/js/systemtags/systemtagmodel.js:
```diff
@@ -39,18 +41,30 @@
 		parse: function(data) {
+			var userEditable;
+			//For invisible and restrict tag set userEditable to false
+			if ((data.userAssignable === false && data.userVisible === false && data.userEditable === true) ||
```
is there a way to simplify this long condition for readability ?

PVince81 commented on core/js/systemtags/systemtagmodel.js:
```diff
@@ -39,18 +41,30 @@
+				(data.userVisible === true && data.userAssignable === false && data.userEditable === true)) {
+				userEditable = false;
+			} else {
+				userEditable = !(data.userEditable === false || data.userEditable === 'false');
```
also include this part in your simplification

the whole thing needs to be more readable

PVince81 commented on core/js/systemtags/systemtags.js:
```diff
@@ -42,6 +42,10 @@
 				scope = t('core', 'invisible');
 			}
+			if (tag.userVisible === true && tag.userEditable === false && tag.userAssignable === true) {
+				// Users cannot edit the tag
```
they can

PVince81 commented on core/js/systemtags/systemtagsinputfield.js:
```diff
@@ -288,6 +288,18 @@
 				this._resultTemplate = Handlebars.compile(RESULT_TEMPLATE);
 			}
+
+			//Do not show the edit if the tag is uneditable
```
"show edit if the tag is editable", please reverse the logic here too for readability

PVince81 commented on db_structure.xml:
```diff
@@ -1099,6 +1108,10 @@
 					<sorting>ascending</sorting>
 				</field>
+				<field>
+					<name>assignable</name>
```
please test updates, I suspect that we'll also need to write a migration to add this column if not exists, in case of update

sharidas replied:
Did a simple test with the migration code added. In version 10.0.10 , created a visibleTag, restrictedTag and invisibleTag. And while upgrading to version 10.0.10.1 ( tar file version ), copied the Migration folder and info.xml of systemtags app to the 10.0.10.1 installed folder. Below are the console log:
```
➜  owncloud4 tar jxf ~/Downloads/owncloud-enterprise-complete-10.0.10.1.tar.bz2                                                                                
mv ow   /* .
➜  owncloud4 mv owncloud/* .
➜  owncloud4 cp -fa ~/test/owncloud3/apps/systemtags/appinfo/Migrations apps/systemtags/appinfo
➜  owncloud4 cp -fa /tmp/backup_stable10010/data .                                             
➜  owncloud4 cp -fa /tmp/backup_stable10010/config.php config                                  
➜  owncloud4 vim apps/systemtags/appinfo/info.xml 
➜  owncloud4 vim apps/systemtags/appinfo/info.xml
➜  owncloud4 ./occ upgrade                                                                                                  
Cannot load Xdebug - it was already loaded
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
2018-10-17T12:35:01+00:00 Set log level to debug
2018-10-17T12:35:01+00:00 Repair step: Repair MySQL database engine
2018-10-17T12:35:01+00:00 Repair info: Not a mysql database -> nothing to do
2018-10-17T12:35:01+00:00 Repair step: Repair MySQL collation
2018-10-17T12:35:01+00:00 Repair info: Not a mysql database -> nothing to do
2018-10-17T12:35:01+00:00 Repair step: Repair SQLite autoincrement
2018-10-17T12:35:01+00:00 Repair step: Repair orphaned reshare
2018-10-17T12:35:01+00:00 Repair step: Repair duplicate entries in oc_lucene_status
2018-10-17T12:35:01+00:00 Repair info: lucene_status table does not exist -> nothing to do
2018-10-17T12:35:01+00:00 Repair step: Upgrade app code from the marketplace
2018-10-17T12:35:01+00:00 Updating database schema
2018-10-17T12:35:01+00:00 Updated database
2018-10-17T12:35:01+00:00 Updating <admin_audit> ...
2018-10-17T12:35:01+00:00 Updated <admin_audit> to 1.0.1
2018-10-17T12:35:01+00:00 Updating <systemtags> ...
2018-10-17T12:35:01+00:00 Updated <systemtags> to 0.3.1
2018-10-17T12:35:01+00:00 Repair step: Repair mime types
2018-10-17T12:35:01+00:00 Repair step: Detect file cache entries with path that does not match parent-child relationships
2018-10-17T12:35:01+00:00 Repair step: Generate ETags for file where no ETag is present.
2018-10-17T12:35:01+00:00 Repair info: ETags have been fixed for 0 files/folders.
2018-10-17T12:35:01+00:00 Repair step: Clean tags and favorites
2018-10-17T12:35:01+00:00 Repair info: 0 tags of deleted users have been removed.
2018-10-17T12:35:01+00:00 Repair info: 0 tags for delete files have been removed.
2018-10-17T12:35:01+00:00 Repair info: 0 tag entries for deleted tags have been removed.
2018-10-17T12:35:01+00:00 Repair info: 0 tags with no entries have been removed.
2018-10-17T12:35:01+00:00 Repair step: Drop old database tables
2018-10-17T12:35:01+00:00 Drop old database tables
2018-10-17T12:35:01+00:00 
                                                    2018-10-17T12:35:01+00:00  Done
 28/28 [============================] 100%2018-10-17T12:35:01+00:00 
2018-10-17T12:35:01+00:00 Repair step: Drop old background jobs
2018-10-17T12:35:01+00:00 Repair step: Remove getetag entries in properties table
2018-10-17T12:35:01+00:00 Repair info: Removed 0 unneeded "{DAV:}getetag" entries from properties table.
2018-10-17T12:35:01+00:00 Repair step: Repair outdated OCS IDs
2018-10-17T12:35:01+00:00 Repair step: Repair invalid shares
2018-10-17T12:35:01+00:00 Repair step: Remove old share propagation app entries
2018-10-17T12:35:01+00:00 Repair step: Move user avatars outside the homes to the new location
2018-10-17T12:35:01+00:00 Repair step: Remove shares of a users root folder
2018-10-17T12:35:01+00:00 Repair step: Repair unmerged shares
2018-10-17T12:35:01+00:00 Repair step: Disable extra themes
2018-10-17T12:35:01+00:00 Repair step: Repair sub shares
2018-10-17T12:35:01+00:00 Starting code integrity check...
2018-10-17T12:35:35+00:00 Finished code integrity check
2018-10-17T12:35:35+00:00 Update successful
2018-10-17T12:35:35+00:00 Maintenance mode is kept active
2018-10-17T12:35:35+00:00 Reset log level
➜  owncloud4  sqlite3 data/owncloud.db
sqlite> select * from oc_systemtag;
id          name        visibility  editable    assignable
----------  ----------  ----------  ----------  ----------
1           visibleTag  1           1           1         
2           restrictTa  1           0           0         
3           invisibleT  0           0           0         
sqlite> 
➜  owncloud4
```

PVince81 commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -431,4 +447,23 @@ public function getTagGroups(ISystemTag $tag) {
+	/**
+	 * {@inheritdoc}
+	 */
+	public function userEditableInGroup(ISystemTag $tag, IUser $user) {
```
`isUserEditableInGroup` ?

PVince81 commented on lib/public/SystemTag/ISystemTagManager.php:
```diff
@@ -160,4 +162,15 @@ public function setTagGroups(ISystemTag $tag, $groupIds);
+
+	/**
+	 * Verify if the user can edit the tag or not in a group. Applicable only for
+	 * uneditedTag
```
please fix "uneditedTag"

PVince81 reviewed (commented):
see comments.
there seem to be leftovers of "uneditable"

PVince81 commented:
hmm strange, so no migration required ? this is unlike in apps as I observed that the activity app does this differently as observed in https://github.com/owncloud/activity/issues/652

@sharidas can you grep through all known migrations and confirm that there is no migration touching this table, for consistency ? then we could move forward with merging your PR.

an alternative plan would be to invest a bit more time and move all the system tags table stuff to migrations (create table in migrations, and have another one that updates it, but is more risky)

sharidas commented:
Updated the unit test which was failing.

sharidas commented:
Updated the unit test which was failing.

sharidas commented:
Updated tests which were failing.

sharidasan commented:
I have moved the Migrations to core ( from systemtags app). Below is the console log of upgrade from 10.0.10 to 10.0.10.1
```
➜  owncloud4 rm -fr *; tar jxf ~/Downloads/owncloud-enterprise-complete-10.0.10.1.tar.bz2; mv owncloud/* .
zsh: sure you want to delete all 28 files in /home/sujith/test/owncloud4 [yn]? y
➜  owncloud4 cp -fa ../backup_upgrade_systemtags/config.php config/; cp -fa ../backup_upgrade_systemtags/data .
➜  owncloud4 cp -fa ~/test/owncloud3/apps/systemtags/appinfo/Migrations/Version201810171* core/Migrations      
➜  owncloud4 vim core/Migrations/Version20181017120818.php 
➜  owncloud4 vim core/Migrations/Version20181017105216.php 
➜  owncloud4 ./occ upgrade
Cannot load Xdebug - it was already loaded
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
2018-10-26T10:59:37+00:00 Set log level to debug
2018-10-26T10:59:37+00:00 Repair step: Repair MySQL database engine
2018-10-26T10:59:37+00:00 Repair info: Not a mysql database -> nothing to do
2018-10-26T10:59:37+00:00 Repair step: Repair MySQL collation
2018-10-26T10:59:37+00:00 Repair info: Not a mysql database -> nothing to do
2018-10-26T10:59:37+00:00 Repair step: Repair SQLite autoincrement
2018-10-26T10:59:37+00:00 Repair step: Repair orphaned reshare
2018-10-26T10:59:37+00:00 Repair step: Repair duplicate entries in oc_lucene_status
2018-10-26T10:59:37+00:00 Repair info: lucene_status table does not exist -> nothing to do
2018-10-26T10:59:37+00:00 Repair step: Upgrade app code from the marketplace
2018-10-26T10:59:37+00:00 Updating database schema
2018-10-26T10:59:37+00:00 Updated database
2018-10-26T10:59:37+00:00 Updating <admin_audit> ...
2018-10-26T10:59:37+00:00 Updated <admin_audit> to 1.0.1
2018-10-26T10:59:37+00:00 Repair step: Repair mime types
2018-10-26T10:59:37+00:00 Repair step: Detect file cache entries with path that does not match parent-child relationships
2018-10-26T10:59:37+00:00 Repair step: Generate ETags for file where no ETag is present.
2018-10-26T10:59:37+00:00 Repair info: ETags have been fixed for 0 files/folders.
2018-10-26T10:59:37+00:00 Repair step: Clean tags and favorites
2018-10-26T10:59:37+00:00 Repair info: 0 tags of deleted users have been removed.
2018-10-26T10:59:37+00:00 Repair info: 0 tags for delete files have been removed.
2018-10-26T10:59:37+00:00 Repair info: 0 tag entries for deleted tags have been removed.
2018-10-26T10:59:37+00:00 Repair info: 0 tags with no entries have been removed.
2018-10-26T10:59:37+00:00 Repair step: Drop old database tables
2018-10-26T10:59:37+00:00 Drop old database tables
2018-10-26T10:59:37+00:00 
                                                    2018-10-26T10:59:37+00:00  Done
 28/28 [============================] 100%2018-10-26T10:59:37+00:00 
2018-10-26T10:59:37+00:00 Repair step: Drop old background jobs
2018-10-26T10:59:37+00:00 Repair step: Remove getetag entries in properties table
2018-10-26T10:59:37+00:00 Repair info: Removed 0 unneeded "{DAV:}getetag" entries from properties table.
2018-10-26T10:59:37+00:00 Repair step: Repair outdated OCS IDs
2018-10-26T10:59:37+00:00 Repair step: Repair invalid shares
2018-10-26T10:59:37+00:00 Repair step: Remove old share propagation app entries
2018-10-26T10:59:37+00:00 Repair step: Move user avatars outside the homes to the new location
2018-10-26T10:59:37+00:00 Repair step: Remove shares of a users root folder
2018-10-26T10:59:37+00:00 Repair step: Repair unmerged shares
2018-10-26T10:59:37+00:00 Repair step: Disable extra themes
2018-10-26T10:59:37+00:00 Repair step: Repair sub shares
2018-10-26T10:59:37+00:00 Starting code integrity check...
2018-10-26T11:00:09+00:00 Finished code integrity check
2018-10-26T11:00:09+00:00 Update successful
2018-10-26T11:00:09+00:00 Maintenance mode is kept active
2018-10-26T11:00:09+00:00 Reset log level
➜  owncloud4
```
The sql table before upgrade:
```
sqlite> select * from oc_systemtag;
id          name        visibility  editable  
----------  ----------  ----------  ----------
1           visibleTag  1           1         
2           restrictTa  1           0         
3           invisibleT  0           0         
sqlite>
```
The sql table after upgrade
```
sqlite> select * from oc_systemtag;
id          name        visibility  editable    assignable
----------  ----------  ----------  ----------  ----------
1           visibleTag  1           1           1         
2           restrictTa  1           0           0         
3           invisibleT  0           0           0         
sqlite>
```

sharidasan commented:
Steps done for verification:
1. In the workflow and tags page under admin -> Settings, the tags are created as below:
![workflowpage](https://user-images.githubusercontent.com/44398014/47565776-3632bb80-d947-11e8-9e5b-81b4ca0fcdf7.gif)
You may also refresh the page to verify all the tags created are there. Its not covered in the screencast shared.
2. Now in my test I have `user1` who belongs to `group1` and `user3` belongs to `group2`. The blow screen cast is from `user3` page:
![user3](https://user-images.githubusercontent.com/44398014/47566209-98d88700-d948-11e8-9e72-9e69ca7bb542.gif)
For `user1` below is the screencast:
![user1](https://user-images.githubusercontent.com/44398014/47566358-2320eb00-d949-11e8-8828-2018a0f07c85.gif)
3. The tag updation from the workflow and tags page is as shown below:
![tagupdate](https://user-images.githubusercontent.com/44398014/47566544-ba863e00-d949-11e8-9544-6b432594c038.gif)

PVince81 commented:
@sharidasan ready for re-review ?

sharidasan commented:
> @sharidasan ready for re-review ?

Yes

PVince81 commented on core/Migrations/Version20181017120818.php:
```diff
@@ -0,0 +1,50 @@
+use OCP\IDBConnection;
+use OCP\Migration\ISqlMigration;
+
+class Version20181017120818 implements ISqlMigration {
```
add comment to explain what we are doing here and why

PVince81 commented on core/js/systemtags/systemtagsinputfield.js:
```diff
@@ -288,6 +288,23 @@
+			 */
+			var editableTag = data.userEditable === false;
+			if (editableTag === true) {
+				/*if (data.editableInGroup === true) {
```
remove commented out code

PVince81 reviewed (dismissed):
See two comments, just for clarification.

Otherwise this looks good to go :+1:

PVince81 commented:
- [ ] adjust acceptance tests for the new flag

sharidas commented:
I have updated the PR. Review required @PVince81

PVince81 commented:
@sharidas I'm missing a PR comment that tells what you actually updated. it just says "updated and retested" with no details. not everyone is going to read the full diff.

sharidas commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -125,7 +127,7 @@ public function getTagsByIds($tagIds) {
 	 * {@inheritdoc}
 	 */
-	public function getAllTags($visibilityFilter = null, $nameSearchPattern = null) {
+	public function getAllTags($visibilityFilter = null, $nameSearchPattern = null, IUser $user = null) {
```
Update: Introduced an argument to this public API of type `IUser`.

sharidas commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -148,11 +150,21 @@ public function getAllTags($visibilityFilter = null, $nameSearchPattern = null)
+			 */
+			if (($user !== null) && $this->isStaticTag($tag) &&
+				!$this->canUserUseStaticTagInGroup($tag, $user)) {
+				continue;
```
Update: This condition would help to remove the static tag which are not to be shown to user. Also added a new private method `isStaticTag` in the same file.

PVince81 replied:
how are invisible tags getting hidden ? there seems to be no condition for this in the method `getAllTags()` which tells me that the intended behavior of the method is to always return everything.
Only if someone passes a "visibilityFilter" value then it would exclude the tags.

Please check the code path for invisible tags as regular users and see if there is another / better location for this check

sharidas commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -431,4 +459,35 @@ public function getTagGroups(ISystemTag $tag) {
+	 * @return bool true if tag is static tag, else false
+	 * @since 10.0.11
+	 */
+	private function isStaticTag(ISystemTag $tag) {
```
A new method added, to check if the tag is static tag or not.

PVince81 replied:
we should rather check for permissions instead of tag type, so instead of querying whether a tag is static, check whether the tag is visible/editable/... depending on context. needs to resolve https://github.com/owncloud/core/pull/33118/files#r229687017 first

sharidas commented on core/js/systemtags/systemtagsinputfield.js:
```diff
@@ -288,6 +289,20 @@
+			 */
+			var editableTag = data.userEditable === false;
+			if (editableTag === true && this._isAdmin === false) {
+				if (data.editableInGroup === true && data.userAssignable === true) {
```
Update: Minor adjustment made to the condition.

sharidas commented on apps/dav/lib/SystemTag/SystemTagsByIdCollection.php:
```diff
@@ -119,7 +119,7 @@ public function getChildren() {
 		}
 
-		$tags = $this->tagManager->getAllTags($visibilityFilter);
+		$tags = $this->tagManager->getAllTags($visibilityFilter, null, $this->userSession->getUser());
```
Updated the call for `getAllTags` based on the change made.

PVince81 commented:
and since you squashed I cannot see the diff between the old behavior and the new one, which makes such summary comment even more important

sharidas commented:
> @sharidas I'm missing a PR comment that tells what you actually updated. it just says "updated and retested" with no details. not everyone is going to read the full diff.

Updated the PR by adding comments on the changes made. I have added `Update` before the comment to make sure they are the updates done. @PVince81

PVince81 commented on core/js/systemtags/systemtagsinputfield.js:
```diff
@@ -288,6 +289,20 @@
+			if (editableTag === true && this._isAdmin === false) {
+				if (data.editableInGroup === true && data.userAssignable === true) {
+					this._allowActions = false;
+				}
```
no else block here ? what is the default ?

PVince81 reviewed (commented):
some concerns about the clarity of the logic and API expectations

PVince81 commented:
@sharidas I guess more work is needed to make the static tag visibility be applied on a different level in the code ?

sharidas commented:
I have updated the code ( old version of the PR ) by removing the change made compared to previous version of PR, where I brought API changes for systemtagmanager's `getAllTags`, also introduced a new method `isStaticTag`. The latest change will get the change from getAllTags and will filter in the UI. Basically modified the `systemtagsinputfield.js` compared to old version of the PR. The reason for modifying this file is, the code in this file shows up how the tags needs to be displayed to the user in UI. Adjusted the conditions in `systemtagsinputfield.js` to show up static tag for the user, also removed the edit options from the static tag. It also removes the delete icon, in the files view for the static tag. 

The static tag with the current PR, works with the conditions discussed:
- Only admin user has the right to edit/remove the static tag.
- When static tag is whitelisted to group(s)
  - Only the users of the group can see the tag. They can only assign or unassign the tag.
  - Users who do not belong to the group cannot see the static tag 
- When static tag does not have any groups whitelisted
  - Only admin user can see, assign/unassign, rename or delete the tag.

Hence the current code works as per the requirement and I would opt for a review. @PVince81

PVince81 commented on core/js/systemtags/systemtagsinputfield.js:
```diff
@@ -288,6 +289,29 @@
+						 * So this is a static tag and user does not have permission
+						 * to see the tag
+						 */
+						return;
```
so what is the value of `this._allowActions` in this one case ?

would be good to reformat / rework the code for better readability

is there any JS unit test code to validate all code paths of this condition ? if not, would be good to add one first before rewriting this

sharidas replied:
The value of `this._allowActions` is `true` in this case. Sure will reformat the the code, as suggested.

PVince81 reviewed (commented):
see comment about readability

also might be good to enhance the JS tests to cover for some of the changes

sharidas commented:
- [x] Reformatted the code in `systemtagsinputfield.js` for better readability
   - Line number 282 - 291 in the same file.
   - Line number 302 - 320 in the same file.
- [x] Updated the unit test for the behaviour to test code updated for `systemtagsinputfield.js`
    - Here is the console log of js test:
```
➜  owncloud3 git:(collaborative-tags-new-category-work) ✗ ./build/node_modules/.bin/karma start tests/karma.config.js 
Coverage preprocessor:  enabled
Apps to test:  [ 'core',
  'files',
  'files_trashbin',
  { name: 'files_sharing',
    srcFiles: 
     [ 'apps/files_sharing/js/app.js',
       'apps/files_sharing/js/sharedfilelist.js',
       'apps/files_sharing/js/share.js',
       'apps/files_sharing/js/external.js',
       'apps/files_sharing/js/public.js',
       'apps/files_sharing/js/sharetabview.js',
       'apps/files_sharing/js/PublicUploadView.js' ],
    testFiles: [ 'apps/files_sharing/tests/js/*.js' ] },
  { name: 'files_external',
    srcFiles: 
     [ 'apps/files_external/js/app.js',
       'apps/files_external/js/mountsfilelist.js',
       'apps/files_external/js/settings.js',
       'apps/files_external/js/statusmanager.js' ],
    testFiles: [ 'apps/files_external/tests/js/*.js' ] },
  { name: 'files_versions',
    srcFiles: 
     [ 'apps/files_versions/js/versionmodel.js',
       'apps/files_versions/js/versioncollection.js',
       'apps/files_versions/js/versionstabview.js' ],
    testFiles: [ 'apps/files_versions/tests/js/**/*.js' ] },
  { name: 'comments',
    srcFiles: 
     [ 'apps/comments/js/app.js',
       'apps/comments/js/commentmodel.js',
       'apps/comments/js/commentcollection.js',
       'apps/comments/js/commentsummarymodel.js',
       'apps/comments/js/commentstabview.js',
       'apps/comments/js/filesplugin.js' ],
    testFiles: [ 'apps/comments/tests/js/**/*.js' ] },
  { name: 'systemtags',
    srcFiles: 
     [ 'apps/systemtags/js/app.js',
       'apps/systemtags/js/systemtagsinfoview.js',
       'apps/systemtags/js/systemtagsfilelist.js',
       'apps/systemtags/js/filesplugin.js' ],
    testFiles: [ 'apps/systemtags/tests/js/**/*.js' ] },
  { name: 'settings',
    srcFiles: 
     [ 'settings/js/users/deleteHandler.js',
       'settings/js/admin-apps.js' ],
    testFiles: [ 'settings/tests/js/apps/appSettingsSpec.js' ] } ]
05 11 2018 15:57:35.805:WARN [watcher]: Pattern "/home/sujith/test/owncloud3/settings/js/users/deleteHandler.js" does not match any file.
05 11 2018 15:57:36.732:WARN [karma]: No captured browser, open http://localhost:9876/
05 11 2018 15:57:36.740:INFO [karma-server]: Karma v3.1.1 server started at http://0.0.0.0:9876/
05 11 2018 15:57:36.740:INFO [launcher]: Launching browsers PhantomJS with concurrency unlimited
05 11 2018 15:57:36.746:INFO [launcher]: Starting browser PhantomJS
05 11 2018 15:57:37.177:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket _EHYEuPYYrSviPAhAAAA with id 16369019
05 11 2018 15:57:37.628:WARN [web-server]: 404: /owncloud/index.php/apps/files_external/userglobalstorages?testOnly=true
05 11 2018 15:57:37.629:WARN [web-server]: 404: /owncloud/index.php/apps/files_external/userstorages
..................................................05 11 2018 15:57:37.777:WARN [web-server]: 404: /svg
................05 11 2018 15:57:37.859:WARN [web-server]: 404: /svg
..............
.......05 11 2018 15:57:37.990:WARN [web-server]: 404: /svg
.............................................................05 11 2018 15:57:38.308:WARN [web-server]: 404: /owncloud/remote.php/dav/avatars/foo/32.jpeg
05 11 2018 15:57:38.345:WARN [web-server]: 404: /owncloud/remote.php/dav/avatars/foo/61.jpeg
05 11 2018 15:57:38.347:WARN [web-server]: 404: /base/core/fonts/OpenSans-Regular.woff
05 11 2018 15:57:38.347:WARN [web-server]: 404: /base/core/fonts/OpenSans-Light.woff
05 11 2018 15:57:38.352:WARN [web-server]: 404: /base/core/fonts/OpenSans-Semibold.woff
............
................................................................................
..........05 11 2018 15:57:38.782:WARN [web-server]: 404: /base/core/css/images/ui-bg_highlight-soft_100_eeeeee_1x100.png
..............................................................05 11 2018 15:57:39.545:WARN [web-server]: 404: /owncloud/remote.php/dav/meta/undefined/v/1437094920.003?preview
........
................................................................................
........................................
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.....
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
............................05 11 2018 15:57:41.143:WARN [web-server]: 404: /base/core/css/images/ui-bg_flat_35_1d2d44_40x100.png
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
...
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
LOG: 'Retrying upload web-file-upload-dd18bf3a8e0a2a3e53e2661c7fb53534-0'
...........05 11 2018 15:57:41.260:WARN [web-server]: 404: /base/core/img/actions/icon.svg
.........05 11 2018 15:57:41.347:WARN [web-server]: 404: /base/core/img/actions/iconAltText.svg
05 11 2018 15:57:41.349:WARN [web-server]: 404: /base/core/img/actions/iconNoAltText.svg
05 11 2018 15:57:41.349:WARN [web-server]: 404: /base/core/img/actions/test.svg
05 11 2018 15:57:41.388:WARN [web-server]: 404: /base/core/img/actions/test.svg
............................................................
...................................05 11 2018 15:57:44.120:WARN [web-server]: 404: /base/core/img/filetypes/folder-alt.svg
.05 11 2018 15:57:44.187:WARN [web-server]: 404: /owncloud/somepath.png
............................................
.........
WARN: 'Missing argument $row in OC.Notification.hide() call, caller needs to be adjusted to only dismiss its own notification'
................................................................................
................................................................................
..05 11 2018 15:57:47.723:WARN [web-server]: 404: /svg

WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
......05 11 2018 15:57:47.785:WARN [web-server]: 404: /svg
........
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
...
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
..
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.......
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
..
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
...
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.
WARN: 'Deprecation warning: tipsy is deprecated. Use tooltip instead.'
.......................05 11 2018 15:57:48.674:WARN [web-server]: 404: /owncloud/remote.php/dav/meta/undefined/v/1437094920.003?preview
.........................................................
.............
PhantomJS 2.1.1 (Linux 0.0.0): Executed 946 of 950 (skipped 4) SUCCESS (11.612 secs / 11.342 secs)
```

PVince81 reviewed (approved):
:+1:

please backport

sharidas commented:
Backported PR : https://github.com/owncloud/core/pull/33420

settermjd commented on core/js/systemtags/systemtagmodel.js:
```diff
@@ -39,18 +41,29 @@
+			if (data.userAssignable === false) {
+				userEditable = false;
+			} else {
+				userEditable = (data.userEditable === true || data.userEditable === 'true');
```
This was confusing to read, at first. Is there not a utility method for this?

sharidas replied:
While developing this feature, I did not found a utility method for this in the frontend.
